service/src/test/java/uk/nhs/adaptors/gp2gp/uat/EhrExtractUATTest.java

```diff
@@ -167,6 +172,49 @@ public void When_MappingValidJsonRequestBody_Expect_ValidXmlOutput(String inputJ
 
         assumeThatCode(() -> validateFileContentAgainstSchema(hl7TranslatedResponse))
             .doesNotThrowAnyException();
+
+        assertThatAgentReferencesAreValid(hl7TranslatedResponse);
+    }
+
+    @SneakyThrows
+    private void assertThatAgentReferencesAreValid(String hl7) {
+        var xPathService = new XPathService();
+        var document = xPathService.parseDocumentFromXml(hl7);
+        var agentRefIdNodes = xPathService.getNodes(document, "//agentRef/id");
+        Set<String> referencedAgentIds = extractIdsFromNodeList(agentRefIdNodes, true);
+        var agentIdNodes = xPathService.getNodes(document, "//Agent/id");
+        Set<String> agentIds = extractIdsFromNodeList(agentIdNodes, false);
+
+        assertThat(referencedAgentIds).isNotEmpty();
+        assertThat(agentIds).isNotEmpty();
+        for (var referencedAgentId : referencedAgentIds) {
+            assertThat(agentIds)
+                .withFailMessage("Expected referenced agent id %s to match the id "
+                    + "of an agent in the agent directory", referencedAgentId)
+                .contains(referencedAgentId);
+        }
```

Comment:
Depending on whether we expect the sets to be equal or one to be a subset of the other, this block could be replaced with `assertThat(agentIds).containsExactlyElementsOf(referencedAgentIds)` or `assertThat(agentIds).containsAll(referencedAgentIds)`. In case of failure, the mismatching IDs would be listed.

This is just a suggestion, up to you.

Reply:
used containsAll since there may be one agent not reference (you always generate one for the patient's managing organisation)